pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,382 @@
+	notifyChannel(sub.name, sub.ch, events, false)
+
+	return sub.ch
+}
+
+// Unsubscribe ends a subscription to entity events and closes its channel.
+func (s *Store) Unsubscribe(ch chan EventBundle) {
+	s.subscribersMut.Lock()
+	defer s.subscribersMut.Unlock()
+
+	for i, sub := range s.subscribers {
+		if sub.ch == ch {
+			s.subscribers = append(s.subscribers[:i], s.subscribers[i+1:]...)
+			break
+		}
+	}
+
+	close(ch)
+}
+
+// GetContainer returns metadata about a container.
+func (s *Store) GetContainer(id string) (Container, error) {
+	var c Container
+
+	entity, err := s.getEntityByKind(KindContainer, id)
+	if err != nil {
+		return c, err
+	}
+
+	c = entity.(Container)
+
+	return c, nil
+}
+
+// GetKubernetesPod returns metadata about a Kubernetes pod.
+func (s *Store) GetKubernetesPod(id string) (KubernetesPod, error) {
+	var p KubernetesPod
+
+	entity, err := s.getEntityByKind(KindKubernetesPod, id)
+	if err != nil {
+		return p, err
+	}
+
+	p = entity.(KubernetesPod)
+
+	return p, nil
+}
+
+// GetECSTask returns metadata about an ECS task.
+func (s *Store) GetECSTask(id string) (ECSTask, error) {
+	var t ECSTask
+
+	entity, err := s.getEntityByKind(KindECSTask, id)
+	if err != nil {
+		return t, err
+	}
+
+	t = entity.(ECSTask)
+
+	return t, nil
+}
+
+// Notify notifies the store with a slice of events.
+func (s *Store) Notify(events []Event) {
+	if len(events) > 0 {
+		s.eventCh <- events
+	}
+}
+
+func (s *Store) startCandidates(ctx context.Context) {
+	// NOTE: s.candidates is not guarded by a mutex as it's only called by
+	// the store itself, and the store runs on a single goroutine
+	for id, c := range s.candidates {
+		err := c.Start(ctx, s)
+
+		// Leave candidates that returned a retriable error to be
+		// re-started in the next tick
+		if err != nil && retry.IsErrWillRetry(err) {
+			log.Debugf("workloadmeta collector %q could not start, but will retry. error: %s", id, err)
+			continue
+		}
+
+		// Store successfully started collectors for future reference
+		if err == nil {
+			log.Infof("workloadmeta collector %q started successfully", id)
+			s.collectors[id] = c
+		} else {
+			log.Info("workloadmeta collector %q could not start. error: %s", id, err)
+		}
+
+		// Remove non-retriable and successfully started collectors
+		// from the list of candidates so they're not retried in the
+		// next tick
+		delete(s.candidates, id)
+	}
+}
+
+func (s *Store) pull(ctx context.Context) {
+	// NOTE: s.collectors is not guarded by a mutex as it's only called by
+	// the store itself, and the store runs on a single goroutine. If this
+	// method is made public in the future, we need to guard it.
+	for id, c := range s.collectors {
+		// Run each pull in its own separate goroutine to reduce
+		// latency and unlock the main goroutine to do other work.
+		go func(id string, c Collector) {
+			err := c.Pull(ctx)
+			if err != nil {
+				log.Warnf("error pulling from collector %q: %s", id, err.Error())
+			}
+		}(id, c)
```

Comment:
do we control the maximum of // goroutine that can be created? 

I think it is not a big deal since we provide a `ctx` timeout. but I'm wondering if a pull of goroutine can be use here. At the same time I'm thinking it can be seen overengineered.